File: cpp/api/src/base/impl/zmq/OutputStreamSocketZmq.cpp
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

#include "OutputStreamSocketZmq.h"
#include "Server.h"
#include "ContextZmq.h"
#include "../../Messages.h"
#include "../../CancelIdGenerator.h"

namespace cameo {

OutputStreamSocketZmq::OutputStreamSocketZmq(const std::string& name) :
	m_name(name),
	m_context(nullptr) {
}

OutputStreamSocketZmq::~OutputStreamSocketZmq() {
	close();
}

void OutputStreamSocketZmq::init(Context * context, const Endpoint& endpoint, RequestSocket * requestSocket) {

	m_context = dynamic_cast<ContextZmq *>(context);

	std::stringstream cancelEndpoint;

	// We define a unique name that depends on the event stream socket object because there can be many (instances).
	cancelEndpoint << "inproc://cancel." << CancelIdGenerator::newId();

	// Create the sockets.
	m_cancelSocket = std::unique_ptr<zmq::socket_t>(new zmq::socket_t(m_context->getContext(), zmq::socket_type::pub));
	m_cancelSocket->bind(cancelEndpoint.str());

	m_socket = std::unique_ptr<zmq::socket_t>(new zmq::socket_t(m_context->getContext(), zmq::socket_type::sub));

	std::vector<std::string> topicsList;

	// Get the topic id.
	std::string topicId = StringId::from(message::Event::STREAM, m_name);

	topicsList.push_back(topicId);
	topicsList.push_back(message::Event::CANCEL);

	for (std::vector<std::string>::const_iterator s = topicsList.begin(); s != topicsList.end(); ++s) {
		m_socket->setsockopt(ZMQ_SUBSCRIBE, s->c_str(), s->length());
	}

	m_socket->connect(endpoint.toString().c_str());
	m_socket->connect(cancelEndpoint.str().c_str());

	// Wait for the connection to be ready.
	// Poll subscriber.
	zmq_pollitem_t items[1];
	items[0].socket = static_cast<void *>(*(m_socket.get()));
	items[0].fd = 0;
	items[0].events = ZMQ_POLLIN;
	items[0].revents = 0;

	while (true) {
		try {
			requestSocket->requestJSON(createSyncStreamRequest(m_name));
		}
		catch (const ConnectionTimeout&) {
			// The server is not accessible.
		}

		// Wait for 100ms.
		int rc = zmq::poll(items, 1, 100);

		// Return when the subscriber received a message.
		if (rc != 0) {
			break;
		}
	}
}

void OutputStreamSocketZmq::send(const std::string& data) {

	zmq::message_t messageData(data.c_str(), data.size());
	m_socket->send(messageData, zmq::send_flags::none);
}

std::string OutputStreamSocketZmq::receive(bool blocking) {

	// Use the message interface.
	zmq::message_t message;

	zmq::recv_flags flags = (blocking ? zmq::recv_flags::none : zmq::recv_flags::dontwait);
	if (m_socket->recv(message, flags).has_value()) {
		// The message exists.
		return std::string(message.data<char>(), message.size());
	}

	return "";
}

void OutputStreamSocketZmq::cancel() {

	if (m_cancelSocket.get() != nullptr) {
		std::string data(message::Event::CANCEL);

		zmq::message_t requestType(data.c_str(), data.length());
		m_cancelSocket->send(requestType, zmq::send_flags::sndmore);

		zmq::message_t requestData(data.c_str(), data.length());
		m_cancelSocket->send(requestData, zmq::send_flags::none);
	}
}

void OutputStreamSocketZmq::close() {
	m_socket->close();
}

}

File: java/api/src/main/java/fr/ill/ics/cameo/base/impl/zmq/OutputStreamSocketZmq.java
package fr.ill.ics.cameo.base.impl.zmq;
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */



import java.util.concurrent.atomic.AtomicBoolean;

import org.json.simple.JSONObject;
import org.json.simple.parser.ParseException;

import fr.ill.ics.cameo.Zmq;
import fr.ill.ics.cameo.base.ConnectionTimeout;
import fr.ill.ics.cameo.base.Context;
import fr.ill.ics.cameo.base.IdGenerator;
import fr.ill.ics.cameo.base.OutputStreamSocket;
import fr.ill.ics.cameo.base.RequestSocket;
import fr.ill.ics.cameo.base.UnexpectedException;
import fr.ill.ics.cameo.base.impl.OutputStreamSocketImpl;
import fr.ill.ics.cameo.messages.JSON;
import fr.ill.ics.cameo.messages.JSON.Parser;
import fr.ill.ics.cameo.messages.Messages;
import fr.ill.ics.cameo.strings.Endpoint;
import fr.ill.ics.cameo.strings.StringId;

public class OutputStreamSocketZmq implements OutputStreamSocketImpl {
	
	private String name;
	private Zmq.Context context;
	private Parser parser;
	private Zmq.Socket subscriberSocket;
	private Zmq.Socket cancelSocket;
	private int applicationId = -1;
	private AtomicBoolean ended = new AtomicBoolean(false);
	private AtomicBoolean canceled = new AtomicBoolean(false);

	public OutputStreamSocketZmq(String name) {
		super();
		this.name = name;
	}
	
	public void init(Context context, Endpoint endpoint, RequestSocket requestSocket, Parser parser) {
		
		this.context = ((ContextZmq)context).getContext();
		this.parser = parser;
		
		// Prepare our context and subscriber
		Zmq.Socket subscriber = this.context.createSocket(Zmq.SUB);
		subscriber.connect(endpoint.toString());
		
		// Subscribe to the topic.
		String topicId = StringId.from(Messages.Event.STREAM, name);
		subscriber.subscribe(topicId);
		
		String cancelEndpoint = "inproc://" + IdGenerator.newStringId();
		
		subscriber.connect(cancelEndpoint);
		subscriber.subscribe(Messages.Event.CANCEL);
		
		Zmq.Socket cancelPublisher = this.context.createSocket(Zmq.PUB);
		cancelPublisher.bind(cancelEndpoint);
		
		// Polling to wait for connection.
		Zmq.Poller poller = this.context.createPoller(1);
		poller.register(subscriber);
		
		while (true) {
			// The server returns a SYNC_STREAM message that is used to synchronize the subscriber.
			try {
				requestSocket.requestJSON(Messages.createSyncStreamRequest(name));
			}
			catch (ConnectionTimeout e) {
				// Do nothing.
			}

			// Return at the first message received by the subscriber.
			poller.poll(100);
			if (poller.pollin(0)) {
				break;
			}
		}
		
		this.subscriberSocket = subscriber;
		this.cancelSocket = cancelPublisher;
	}
	
	/**
	 * Sets the application id.
	 * @param id
	 */
	public void setApplicationId(int id) {
		this.applicationId = id;
	}
	
	public OutputStreamSocket.Output receive()	{
		
		// Loop on recvStr() because in case of configuration multiple=yes, messages can come from different instances.
		while (true) {
			String messageType = this.subscriberSocket.recvStr();
			
			// Cancel can only come from this instance.
			if (messageType.equals(Messages.Event.CANCEL)) {
				canceled.set(true);
				return null;
			}
			
			// Get the second part of the message.
			byte[] messageValue = this.subscriberSocket.recv();

			try {
				// Get the JSON object.
				JSONObject jsonMessage = parser.parse(Messages.parseString(messageValue));
				
				int type = JSON.getInt(jsonMessage, Messages.TYPE);
				
				// Continue if type of message is SYNC_STREAM. Theses messages are only used for the poller.
				if (type == Messages.SYNC_STREAM) {
					continue;
				}
				
				int id = JSON.getInt(jsonMessage, Messages.ApplicationStream.ID);
				
				// Filter on the application id so that only the messages concerning the instance applicationId are processed.
				// Others are ignored.
				if (applicationId == -1 || applicationId == id) {
					
					// Terminate the stream if type of message is STREAM_END.
					if (type == Messages.STREAM_END) {
						ended.set(true);
						return null;
					}
					
					// Here the type of message is STREAM.
					String line = JSON.getString(jsonMessage, Messages.ApplicationStream.MESSAGE);
					boolean endOfLine = JSON.getBoolean(jsonMessage, Messages.ApplicationStream.EOL);
					
					return new OutputStreamSocket.Output(id, line, endOfLine);
				}
				
				// Here, the application id is different from id, then re-iterate.
			}
			catch (ParseException e) {
				throw new UnexpectedException("Cannot parse response: " + messageValue);
			}
		}
	}
	
	public boolean hasEnded() {
		return ended.get();
	}
	
	public void cancel() {
		cancelSocket.sendMore(Messages.Event.CANCEL);
		cancelSocket.send(Messages.Event.CANCEL);
	}
	
	public boolean isCanceled() {
		return canceled.get();
	}
	
	public void terminate() {
		context.destroySocket(subscriberSocket);
	}
}